### ISSUE TITLE
Suggestion: Tone.Offline might benefit from ability to control when the rendering process starts

### ISSUE THREAD

**Issue by gdenisov** [Tonejs/Tone.js#368]
When using Sampler and Players in the Tone.Offline callback I am running into an issue that by the time Sampler or Player nodes have been loaded rendering is already done. It might be useful for offline callback to look something like this:

```
Tone.Offline((Transport, startRender) => {
  new Sampler({
    "C3" : "path/to/C3.mp3",
    "D#3" : "path/to/Dsharp3.mp3",
    "F#3" : "path/to/Fsharp3.mp3",
    "A3" : "path/to/A3.mp3",
   }, () => {
    // ... schedule events
    startRender()
  })
})
```

Of course we could preload buffers outside of Tone.Offline call and then pass them in to the sampler in the callback, but this becomes increasingly more difficult when the node creation is abstracted in another external function

**Comment by tambien** [comment]
Thanks for the suggestion, i like your proposed solution. 

There is currently an undocumented way to do this, but i'm not totally satisfied with the API yet. Currently you can return a promise within Tone.Offline, and the context will wait until that promise is resolved before starting to render.

**Comment by rodw** [comment]
@tambien can you give an example of what you mean by "return a promise within Tone.Offline, and the context will wait until that promise is resolved before starting to render"?

**Comment by rodw** [comment]
The attached HTML document (added with a `.txt` extension because GitHub won't let me upload it otherwise) demonstrates the problem, or at least the way I'm trying to invoke this.

When opened, the HTML document has four links:

- Basic Synth (online) - plays some notes using Tone.Synth directly
- Sampler (online) - plays some notes using Tone.Sampler directly
- Basic Synth (offline) - plays some notes using Tone.Synth via Tone.Offline (then plays the generated buffer)
- Sampler (offine) - plays some notes using Tone.Sampler via Tone.Offline (then plays the generated buffer), or at least tries to
 
The first three links seem to work fine and can be repeated an arbitrary number of times.

The fourth (offline sampler) doesn't work as-is and I'm not able to get it to work in any variation I've tried.  In the current setup I'm getting "Error connecting to node: [object GainNode]" when I try to call `sampler.triggerAttackRelease`.

That error (and some other debugging output) can be seen in the console log when that link is hit. 

@tambien Am I doing this incorrectly? Should I open a new issue with this example? 

[offline-sampler.html.txt](https://github.com/Tonejs/Tone.js/files/2506792/offline-sampler.html.txt)

**Comment by gdenisov** [comment]
@rodw I have not tried to use this yet but I think you should be able to do something like this:

```


(See https://github.com/Tonejs/Tone.js/blob/092ffc60a14f317485a2b5285ef659647e392647/Tone/core/Offline.js#L71 for reference)

**Comment by gdenisov** [comment]
@rodw I have not tried running this, but it seems that the offline example with the promise is missing transport start. Try adding `transport.start(0.2)` near `resolve()` in `play_via_offline_sampler`

**Comment by rodw** [comment]
@gdenisov thanks for your help, but I don't think it makes much of a difference.

Adding `transport.start(0)` before or after the call to `resolve()` doesn't seem to help.  
 
I have noticed that adding a `synth.sync()` call into the mix (e.g., within `after_loaded`) will make the "Error connecting to node" exception go away, but still doesn't produce any audible output.

In fact no combination of the methods `transport.stop`, `transport.start` or `synth.sync` in or around that `after_loaded` callback seems to work.

For what is is worth, the non-promise, Tone.Synth version works without any calls to `transport.start` (or `synth.sync` or `transport.stop`), but also works with them.

Because it is easier to read, here's the non-promise, Tone.Synth version:

```
      function play_via_offline_synth() {
        var generate_offline = function(transport) {
          transport.stop(); // <== optional line
          var synth = new Tone.Synth().toMaster();
          synth.sync(); // <== optional line
          play_riff(synth);
          transport.start(); // o<== ptional line
        }
        Tone.Offline(generate_offline, 4).catch(console.error).then(function(buffer){
          play_audio_buffer(buffer._buffer);
        });
      }
```

Any combination of including or not including the lines marked `// optional line` seems to work OK.

Putting any combination of those lines in equivalent locations within `play_via_offline_sampler` doesn't seem to help though, at least not in any of the several variations I have tried.

**Comment by gdenisov** [comment]
@rodw Here is the snippet that works for me (which prompted this issue). I am not sure why this works but the promise resolution does not yet. Will need to do some research. Either way if you replace your sample `play_via_offline_synth` with this:

```

      function renderOffline(callback, duration){
        // Set the OfflineAudioContext
        let sampleRate = Tone.context.sampleRate
        let originalContext = Tone.context
        // Make sure to preserve original Transport if it has not be cached on the context yet
        if (!originalContext.Transport) originalContext.Transport = Tone.Transport
        // Create new OfflineContext and make it the default context
        let context = new Tone.OfflineContext(2, duration, sampleRate)
        Tone.context = context
        function onReady() {
          // process the audio
          var rendered = context.render()
          // return the original AudioContext
          Tone.context = originalContext
          // return the audio
          return rendered.then(function(buffer){
            // wrap it in a Tone.Buffer
            return new Tone.Buffer(buffer)
          })
        }
        // invoke the callback/scheduling
        callback(Tone.Transport, onReady)
      }

      function play_via_offline_sampler() {
        renderOffline((Transport, render) => {
          const synth = new Tone.Sampler(samples, onSynthLoaded).toMaster();

          function onSynthLoaded() {
            play_riff(synth);
            render().then((buffer) => {
              buffer.reverse()
              play_audio_buffer(buffer._buffer)
            })
          }
        }, 4)
      }

**Comment by gdenisov** [comment]
@rodw Here is the above as stand alone pen: https://codepen.io/anon/pen/BqqwXK

**Comment by rodw** [comment]
@gdenisov Thanks!  So that seems to work for me also.

If I understand your `renderOffline` method correctly you have essentially re-written the Tone.Offline functionality to match the method signature and behavior you described in the original post for this issue, is that correct?

For what it's worth I actually prefer this syntax to the "return a promise" style that should work according to the @tambien 's comment above and, as far as I can tell, from looking at the `Tone.Offline` code itself, but I'm maybe old-school and prefer callbacks to promises in general.  (I've resigned myself to this being a minority opinion, so don't read that as advocating for callbacks rather than promises in Tone.js in general).

Thanks again @gdenisov, I think this solves my issue.

Thanks also @tambien for a pretty awesome library in general.  My issues with offline-rendering notwithstanding I'm impressed and pleasantly surprised at how easy Tone.js makes working with synths and samplers _and_ how easy it is to mix Tone.js with direct use of the WebAudio API.

**Comment by tambien** [comment]
Thanks for all your investigations and comments. The current API is clunky at the moment and was meant as a temporary bandaid until i can address the problem better. 

I think other than the API being undocumented and confusing, the underlying issue is that the context AudioContext is not captured in the function's closure, but instead set globally only while the function inside of Tone.Offline is being invoked. so any deferred callback that's not in the initial callback will not have the same AudioContext, meaning that it will not be recorded and that it can't be connected to the first OfflineAudioContext. 

The more sustainable solution would be to enable a non-temporary way to set the context within the callback, which is a bit complicated, but i've made a little progress on

**Comment by teropa** [comment]
> The more sustainable solution would be to enable a non-temporary way to set the context within the callback, which is a bit complicated, but i've made a little progress on

Not sure how you'd feel about introducing a dependency, but this seems like the kind of problem [zone.js](https://github.com/angular/zone.js) was designed to address. 

I.e. Within `Tone.Offline` you'd be in an "offline zone" which would then persist throughout any async work launched from that function, no matter what kind.

**Comment by tambien** [comment]
i'm going to close this. this is pretty much solved with the latest `tone@next` which allows you to pass in a `context` to every node which enables you to control exactly when and how the rendering happens using Tone.OfflineContext. 

example:

```javascript
const offline = new Tone.OfflineContext(2, 30, 44100)
const oscillator = new Oscillator({context: offline})
// ... async
const buffer = await offline.render()
```

**Comment by cristiano-belloni** [comment]
@tambien, moving to the latest `tone@next` breaks my whole application. Did you change anything major? I can't even import it:

`import * as Tone from "tone"`

`'tone' does not contain a default export (imported as 'Tone').`
